counsel.el

```diff
@@ -820,14 +834,18 @@ Optional INITIAL-INPUT is the initial input in the minibuffer."
   (let* ((cands obarray)
          (pred 'commandp)
          (sort t))
-    (when (require 'smex nil 'noerror)
-      (unless smex-initialized-p
-        (smex-initialize))
-      (when (smex-detect-new-commands)
-        (smex-update))
-      (setq cands smex-ido-cache)
-      (setq pred nil)
-      (setq sort nil))
+    (cond (smex-initialized-p
+           (when (smex-detect-new-commands)
+             (smex-update))
+           (setq cands smex-ido-cache)
+           (setq pred nil)
+           (setq sort nil))
+          (amx-initialized
+           (when (amx-detect-new-commands)
+             (amx-update))
+           (setq cands amx-cache)
+           (setq pred nil)
+           (setq sort nil)))
```

Comment:
I think this code duplication warrants a separate function which returns either the candidates from `smex`, those from `amx`, or `nil`.